### ISSUE TITLE
merge_asof hangs indefinitely

### ISSUE THREAD

**Issue by oscarknagg** [modin-project/modin#4532]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: 20.04.4 LTS (Focal Fossa)
- **Modin version** (`modin.__version__`): 0.14.1
- **Python version**: Python 3.8.10
- **Code we can use to reproduce**:
```python
import modin.pandas as pd
import pandas as _pd
import numpy as np
import ray
from string import ascii_uppercase
from typing import List
import time


def generate_left_series(by_keys: List[str], n: int, t: float):
    frames = []
    for k in by_keys:
        frames.append(
            _pd.DataFrame({
                "time": np.linspace(0, t, n),
                "by_key": k,
                "value_left": np.random.randn(n)
            })
        )

    return _pd.concat(frames)


def generate_right_series(by_keys: List[str], n: int, t: float):
    frames = []
    for k in by_keys:
        frames.append(
            _pd.DataFrame({
                "time": np.linspace(0, t, n),
                "by_key": k,
                "value_right": np.random.uniform(-1, 1, size=n)
            })
        )

    return _pd.concat(frames)


def random_string(l: int):
    return "".join(np.random.choice(uppercase_letters, size=l))


if __name__ == "__main__":
    ray.init()
    print(ray.cluster_resources())

    uppercase_letters = list(ascii_uppercase)

    n_by_keys = 100
    n_left = 100000
    n_right = 2317

    by_keys = [random_string(4) for _ in range(n_by_keys)]

    left = generate_left_series(by_keys, n_left, float(n_left))
    left = left.sort_values(by=["time"])

    right = generate_right_series(by_keys, n_right, float(n_left))
    right = right.sort_values(by=["time"])

    left_modin = pd.DataFrame(left)
    right_modin = pd.DataFrame(right)

    t0 = time.time()
    _pd.merge_asof(left, right, on="time", by="by_key")
    t = time.time()
    print("Merged pandas dataframes in {:.2f}s".format(t-t0))

    t0 = time.time()
    pd.merge_asof(left_modin, right_modin, on="time", by="by_key")
    t = time.time()
    print("Merged modin dataframes in {:.2f}s".format(t-t0))

```

### Describe the problem

Doing a merge_asof of two small dataframes hangs for at least 5 minutes. The pandas implementation finishes in a few seconds. 

### Source code / logs

If I cancel the program then I get the following stacktrace.
```
Traceback (most recent call last):
  File "merge_asof_hangs.py", line 71, in <module>
    pd.merge_asof(left_modin, right_modin, on="time", by="by_key")
  File "/home/oscark@ilab.local/PycharmProjects/tig-team-oscar/projects/modin-eval/venv/lib/python3.8/site-packages/modin/pandas/general.py", line 279, in merge_asof
    right_subset = right.reindex(index=pandas.Index(merged["right_labels"]))
  File "/home/oscark@ilab.local/PycharmProjects/tig-team-oscar/projects/modin-eval/venv/lib/python3.8/site-packages/modin/pandas/dataframe.py", line 1716, in reindex
    return super(DataFrame, self).reindex(
  File "/home/oscark@ilab.local/PycharmProjects/tig-team-oscar/projects/modin-eval/venv/lib/python3.8/site-packages/modin/pandas/base.py", line 2241, in reindex
    new_query_compiler = self._query_compiler.reindex(
  File "/home/oscark@ilab.local/PycharmProjects/tig-team-oscar/projects/modin-eval/venv/lib/python3.8/site-packages/modin/core/storage_formats/pandas/query_compiler.py", line 515, in reindex
    new_modin_frame = self._modin_frame.apply_full_axis(
  File "/home/oscark@ilab.local/PycharmProjects/tig-team-oscar/projects/modin-eval/venv/lib/python3.8/site-packages/modin/core/dataframe/pandas/dataframe/dataframe.py", line 114, in run_f_on_minimally_updated_metadata
    result = f(self, *args, **kwargs)
  File "/home/oscark@ilab.local/PycharmProjects/tig-team-oscar/projects/modin-eval/venv/lib/python3.8/site-packages/modin/core/dataframe/pandas/dataframe/dataframe.py", line 1897, in apply_full_axis
    return self.broadcast_apply_full_axis(
  File "/home/oscark@ilab.local/PycharmProjects/tig-team-oscar/projects/modin-eval/venv/lib/python3.8/site-packages/modin/core/dataframe/pandas/dataframe/dataframe.py", line 114, in run_f_on_minimally_updated_metadata
    result = f(self, *args, **kwargs)
  File "/home/oscark@ilab.local/PycharmProjects/tig-team-oscar/projects/modin-eval/venv/lib/python3.8/site-packages/modin/core/dataframe/pandas/dataframe/dataframe.py", line 2336, in broadcast_apply_full_axis
    result = self.__constructor__(
  File "/home/oscark@ilab.local/PycharmProjects/tig-team-oscar/projects/modin-eval/venv/lib/python3.8/site-packages/modin/core/dataframe/pandas/dataframe/dataframe.py", line 213, in __init__
    self._filter_empties()
  File "/home/oscark@ilab.local/PycharmProjects/tig-team-oscar/projects/modin-eval/venv/lib/python3.8/site-packages/modin/core/dataframe/pandas/dataframe/dataframe.py", line 433, in _filter_empties
    [
  File "/home/oscark@ilab.local/PycharmProjects/tig-team-oscar/projects/modin-eval/venv/lib/python3.8/site-packages/modin/core/dataframe/pandas/dataframe/dataframe.py", line 440, in <listcomp>
    if i < len(self._row_lengths) and self._row_lengths[i] != 0
  File "/home/oscark@ilab.local/PycharmProjects/tig-team-oscar/projects/modin-eval/venv/lib/python3.8/site-packages/modin/core/dataframe/pandas/dataframe/dataframe.py", line 227, in _row_lengths
    self._row_lengths_cache = [
  File "/home/oscark@ilab.local/PycharmProjects/tig-team-oscar/projects/modin-eval/venv/lib/python3.8/site-packages/modin/core/dataframe/pandas/dataframe/dataframe.py", line 228, in <listcomp>
    obj.length() for obj in self._partitions.T[0]
  File "/home/oscark@ilab.local/PycharmProjects/tig-team-oscar/projects/modin-eval/venv/lib/python3.8/site-packages/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py", line 265, in length
    self._length_cache = ray.get(self._length_cache)
  File "/home/oscark@ilab.local/PycharmProjects/tig-team-oscar/projects/modin-eval/venv/lib/python3.8/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/home/oscark@ilab.local/PycharmProjects/tig-team-oscar/projects/modin-eval/venv/lib/python3.8/site-packages/ray/worker.py", line 1803, in get
    values, debugger_breakpoint = worker.get_objects(object_refs, timeout=timeout)
  File "/home/oscark@ilab.local/PycharmProjects/tig-team-oscar/projects/modin-eval/venv/lib/python3.8/site-packages/ray/worker.py", line 362, in get_objects
    data_metadata_pairs = self.core_worker.get_objects(
  File "python/ray/_raylet.pyx", line 1198, in ray._raylet.CoreWorker.get_objects
  File "python/ray/_raylet.pyx", line 167, in ray._raylet.check_status
KeyboardInterrupt
```

**Comment by mvashishtha** [comment]
@oscarknagg thank you for reporting this issue! I can reproduce the behavior on my mac, where pandas takes 2.45 seconds and modin has not finished after 40 minutes. On the Ray dashboard, I see that a single task has about 100% utilization and is using 2-3 GB of RAM (the usage varies). I'm investigating this some more. Meanwhile, note that `merge_asof` is defaulting to pandas, so we don't expect modin to give better performance that pandas.

**Comment by mvashishtha** [comment]
I tracked the bottleneck to a slow reindex in the Modin code: https://github.com/modin-project/modin/blob/dcee13d57ebf9a006460deedb734c15791acae7a/modin/pandas/general.py#L285

reindexing the much smaller right frame (length 231_700) with the 40x larger merged index (length 10_000_000) takes a very, very long time. Note that it seems we need the indices to contain lots of duplicates for the reindex to be slow. If I give each frame the default `RangeIndex`, Modin only takes a few seconds for the merge, too.

Here's a shorter script that reproduces the bug in Modin:

```python
import modin.pandas as pd
import pandas
import time

pandas_left = pandas.DataFrame({'x': list(range(100_000 * 100))}, index=list(range(100_000)) * 100)
pandas_right = pandas.DataFrame({'x': list(range(2317 * 100))}, index=list(range(2317)) * 100)
start = time.time()
pandas.merge_asof(pandas_left, pandas_right, on='x')
end = time.time()
print(f'pandas merge took {end - start}')

modin_left = pd.DataFrame(pandas_left)
modin_right = pd.DataFrame(pandas_right)
print('starting modin merge...')
pd.merge_asof(modin_left, modin_right, on='x')
```

I'm checking whether we can avoid this reindex or skip some of the work when there are duplicates.

**Comment by mvashishtha** [comment]
@oscarknagg it turns out that not only is Modin way too slow here, but also it would eventually raise an error due to #4541 😢. I have a draft fix #4542 for that bug. The fix makes Modin take 4.48 seconds for this script. modin does worse than pandas even with the fix, but again we don't expect good performance because we don't have a distributed `merge_asof` implementation. 

For now, please try out the fix with `pip install git+https://github.com/mvashishtha/modin.git@4541-merge-asof-non-unique-right-index`. Let us know here whether modin's performance with the fix is good enough for you.

**Comment by mvashishtha** [comment]
@oscarknagg given that #4541 is closed, I'm going to close this issue too.
